### ISSUE TITLE
Is there any hook example?

### ISSUE THREAD

**Issue by xkguq007** [RobinBuschmann/sequelize-typescript#13]
I searched, but i can't see.

here is my sample table

```
@Table({
  indexes: [],
  instanceMethods: {},
  classMethods: {},
  timestamps: true
})
export class Role extends Model<Role> {
  @PrimaryKey
  @Column(DataType.UUID)
  id: string;

  @BelongsToMany(() => Group, () => RoleGroup)
  groups: Group[];

  @CreatedAt
  createAt: Date;

  @UpdatedAt
  updateAt: Date;
}

Role.beforeCreate(function(model, options): Promise<any> {
  return new Promise((resolve, reject) => {
      model.myInstanceMethodsSample(model.myFieldSample);
      resolve();
  });
});
```

My question is... I don't know, model equals above Role class. because as you can see,
I need to use instance method in this beforeCreate hook.

It will be great that i can see any example in hook.

p.s.
In decorator Table, there is classmethods and instancemethods. but it is useless because I just write function in class. am I right?

**Comment by xkguq007** [comment]
I've been done changing instance methods to class methods.
and, call like Role.classMethods(model.getDataValue('field')) this.
but, is there any smooth way?
this way feel wierd because this is hook function but I use class methods calling like CLASS.FUNCTION
I think this or model object or something else that i can access class methods or instance methods.

**Comment by RobinBuschmann** [comment]
Hey, thanks for contributing.

The `classMethods` and `instanceMethods` property comes from native sequelize. You're right, in `sequelize-typescript` you should use real instance methods and static methods like:

````typescript
@Table
class User {
  
  @Column
  name: string;

  @Column
  reversedName: string;

  prepareReversedName() {

    this.reversedName = this.name.split('').reverse().join('');
  }

  static createSpecialUser(values) {
    / * ... */
  }
}
````

So you want to call an instance method in your hook, am I right? This should work as you already defined in your example. 
````typescript
User.beforeCreate(user => {
   user.prepareReversedName();
});
````
Please note, that you don't need to use a Promise, if nothing async happens. Otherwise if `myInstanceMethodsSample` is async, you should consider the asynchronous operation, what you didn't do in your example. 

But I'm not really sure, whether I really understand, what your actual problem is. Can you give me an example of what you're trying to achieve?

**Comment by zlokomatic** [comment]
Hi,

i tried using a hook:

User.beforeCreate((user : User) => {
    console.log('BEFORE', user);
})

but im getting the error:

```
this.options.hooks[hookType] = this.options.hooks[hookType] || [];
                ^

TypeError: Cannot read property 'hooks' of undefined
```

any hint what im doing wrong?

**Comment by RobinBuschmann** [comment]
@zlokomatic Ah, I see. Calling static members of a model is currently only available after these models are initialized by sequelize. So you need to add these models to the sequelize instance before you can call `beforeCreate`:

````typescript

@Table
class User extends Model<User> {}

const sequelize =  new Sequelize({
        ...,
        modelPaths: ['path/to/models/dir']
});

// or
sequelize.addModels([User]);

User.beforeCreate(...);
User.findAll(...);

````

Hmm, this should be added to the documentation, I think. Thanks raising this issue.

**Comment by xkguq007** [comment]
Oh, then it just tslint alert..
Yeah I got a same problem with @zlokomatic 's problem.
I was trying hundreds of times, so I did like @RobinBuschmann 's did. :)

I just use this library as a test, before I change my backend as Node.js typescript.
This library is really good for begins with Object Oriented style.
Of course I should do more testing, but this project and https://github.com/evertonrobertoauler/universal-demo this project is really good basic for node.js, angular4, sequelize with typescript.

I close this issue.

p.s. sorry for terrible english
